konnectivity-socks5-proxy/main.go

```diff
@@ -0,0 +1,173 @@
+			panic(err)
+		}
+	}
+
+	return cmd
+}
+
+func dialKonnectivityFunc(caCertPath string, clientCertPath string, clientKeyPath string, proxyHostname string, proxyPort int) func(ctx context.Context, network string, addr string) (net.Conn, error) {
+	return func(ctx context.Context, network string, addr string) (net.Conn, error) {
+		caCert := caCertPath
+		tlsConfig, err := util.GetClientTLSConfig(caCert, clientCertPath, clientKeyPath, proxyHostname, nil)
+		if err != nil {
+			return nil, err
+		}
+		var proxyConn net.Conn
+
+		proxyAddress := fmt.Sprintf("%s:%d", proxyHostname, proxyPort)
+		requestAddress := addr
+
+		proxyConn, err = tls.Dial("tcp", proxyAddress, tlsConfig)
+		if err != nil {
+			return nil, fmt.Errorf("dialing proxy %q failed: %v", proxyAddress, err)
+		}
+		fmt.Fprintf(proxyConn, "CONNECT %s HTTP/1.1\r\nHost: %s\r\n\r\n", requestAddress, "127.0.0.1")
+		br := bufio.NewReader(proxyConn)
+		res, err := http.ReadResponse(br, nil)
+		if err != nil {
+			return nil, fmt.Errorf("reading HTTP response from CONNECT to %s via proxy %s failed: %v",
+				requestAddress, proxyAddress, err)
+		}
+		if res.StatusCode != 200 {
+			return nil, fmt.Errorf("proxy error from %s while dialing %s: %v", proxyAddress, requestAddress, res.Status)
+		}
+
+		// It's safe to discard the bufio.Reader here and return the
+		// original TCP conn directly because we only use this for
+		// TLS, and in TLS the client speaks first, so we know there's
+		// no unbuffered data. But we can double-check.
+		if br.Buffered() > 0 {
+			return nil, fmt.Errorf("unexpected %d bytes of buffered data from CONNECT proxy %q",
+				br.Buffered(), proxyAddress)
+		}
+		return proxyConn, nil
+	}
+}
+
+// K8sServiceResolver attempts to resolve the hostname by matching it to a Kubernetes Service, but will fallback to theystem DNS if an error is encountered.
+type K8sServiceResolver struct {
+	client client.Client
+}
+
+func (d K8sServiceResolver) Resolve(ctx context.Context, name string) (context.Context, net.IP, error) {
+	_, ip, err := d.ResolveK8sService(ctx, name)
+	if err != nil {
+		fmt.Printf("Error resolving k8s service %v, falling back to system DNS\n", err)
+		return socks5.DNSResolver{}.Resolve(ctx, name)
+	}
+
+	return ctx, ip, nil
+}
+
+func (d K8sServiceResolver) ResolveK8sService(ctx context.Context, name string) (context.Context, net.IP, error) {
+	NamespaceNamedService := strings.Split(name, ".")
+	if len(NamespaceNamedService) < 2 {
+		return nil, nil, fmt.Errorf("unable to derrive namespacedName from %v", name)
+	}
+	namespacedName := types.NamespacedName{
+		Namespace: NamespaceNamedService[1],
+		Name:      NamespaceNamedService[0],
+	}
+
+	service := &corev1.Service{}
+	err := d.client.Get(ctx, namespacedName, service)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Convert service name to ip address...
+	ip := net.ParseIP(service.Spec.ClusterIP)
+	if ip == nil {
+		return nil, nil, fmt.Errorf("unable to parse IP %v", name)
```

Comment:
probably should include the IP that you couldn't parse (`service.Spec.ClusterIP`)

Reply:
Agreed - I've replaced the `name` variable with the `ip` variable.